File: packages/runtime-client-ts/src/domains/kanban.ts
import type { components } from "../generated/schema";
import type { RuntimeTransport } from "../transport/types";

function enc(id: string): string {
  return encodeURIComponent(id);
}

export type KanbanCapabilities = components["schemas"]["KanbanCapabilities"];
export type KanbanBoard = components["schemas"]["KanbanBoard"];
export type KanbanBoardListResponse = components["schemas"]["KanbanBoardListResponse"];
export type CreateKanbanBoardInput = components["schemas"]["CreateKanbanBoardInput"];
export type KanbanTask = components["schemas"]["KanbanTask"];
export type KanbanTaskListResponse = components["schemas"]["KanbanTaskListResponse"];
export type CreateKanbanTaskInput = components["schemas"]["CreateKanbanTaskInput"];
export type KanbanTaskDetail = components["schemas"]["KanbanTaskDetail"];
export type KanbanTaskActionInput = components["schemas"]["KanbanTaskActionInput"];
export type KanbanComment = components["schemas"]["KanbanComment"];
export type KanbanCommentCreate = components["schemas"]["KanbanCommentCreate"];
export type KanbanEvent = components["schemas"]["KanbanEvent"];
export type KanbanRun = components["schemas"]["KanbanRun"];
export type KanbanAssignee = components["schemas"]["KanbanAssignee"];
export type KanbanAssigneeListResponse = components["schemas"]["KanbanAssigneeListResponse"];
export type KanbanDispatchRequest = components["schemas"]["KanbanDispatchRequest"];
export type KanbanDispatchResult = components["schemas"]["KanbanDispatchResult"];

export interface KanbanTaskFilter {
  status?: string;
  assignee?: string;
  tenant?: string;
  includeArchived?: boolean;
}

export interface KanbanDomain {
  getCapabilities(instanceId: string, signal?: AbortSignal): Promise<KanbanCapabilities>;
  listBoards(
    instanceId: string,
    opts?: { includeArchived?: boolean },
    signal?: AbortSignal,
  ): Promise<KanbanBoardListResponse>;
  createBoard(
    instanceId: string,
    body: CreateKanbanBoardInput,
    signal?: AbortSignal,
  ): Promise<KanbanBoard>;
  removeBoard(instanceId: string, boardSlug: string, signal?: AbortSignal): Promise<void>;
  listTasks(
    instanceId: string,
    boardSlug: string,
    filter?: KanbanTaskFilter,
    signal?: AbortSignal,
  ): Promise<KanbanTaskListResponse>;
  getTask(
    instanceId: string,
    boardSlug: string,
    taskId: string,
    signal?: AbortSignal,
  ): Promise<KanbanTaskDetail>;
  createTask(
    instanceId: string,
    boardSlug: string,
    body: CreateKanbanTaskInput,
    signal?: AbortSignal,
  ): Promise<KanbanTask>;
  executeTaskAction(
    instanceId: string,
    boardSlug: string,
    taskId: string,
    body: KanbanTaskActionInput,
    signal?: AbortSignal,
  ): Promise<KanbanTask>;
  addComment(
    instanceId: string,
    boardSlug: string,
    taskId: string,
    text: string,
    signal?: AbortSignal,
  ): Promise<void>;
  listAssignees(
    instanceId: string,
    boardSlug: string,
    signal?: AbortSignal,
  ): Promise<KanbanAssigneeListResponse>;
  dispatch(
    instanceId: string,
    boardSlug: string,
    body?: KanbanDispatchRequest,
    signal?: AbortSignal,
  ): Promise<KanbanDispatchResult>;
}

export function createKanbanDomain(transport: RuntimeTransport): KanbanDomain {
  const base = (instanceId: string) => `/api/v1/instances/${enc(instanceId)}/kanban`;

  return {
    getCapabilities: (instanceId, signal) =>
      transport.request<KanbanCapabilities>({
        path: `${base(instanceId)}/capabilities`,
        signal,
      }),
    listBoards: (instanceId, opts, signal) =>
      transport.request<KanbanBoardListResponse>({
        path: `${base(instanceId)}/boards`,
        query: opts?.includeArchived ? { includeArchived: "true" } : undefined,
        signal,
      }),
    createBoard: (instanceId, body, signal) =>
      transport.request<KanbanBoard>({
        method: "POST",
        path: `${base(instanceId)}/boards`,
        body,
        signal,
      }),
    removeBoard: async (instanceId, boardSlug, signal) => {
      await transport.request({
        method: "DELETE",
        path: `${base(instanceId)}/boards/${enc(boardSlug)}`,
        signal,
        parseJson: false,
      });
    },
    listTasks: (instanceId, boardSlug, filter, signal) =>
      transport.request<KanbanTaskListResponse>({
        path: `${base(instanceId)}/boards/${enc(boardSlug)}/tasks`,
        query: {
          status: filter?.status,
          assignee: filter?.assignee,
          tenant: filter?.tenant,
          includeArchived: filter?.includeArchived ? "true" : undefined,
        },
        signal,
      }),
    getTask: (instanceId, boardSlug, taskId, signal) =>
      transport.request<KanbanTaskDetail>({
        path: `${base(instanceId)}/boards/${enc(boardSlug)}/tasks/${enc(taskId)}`,
        signal,
      }),
    createTask: (instanceId, boardSlug, body, signal) =>
      transport.request<KanbanTask>({
        method: "POST",
        path: `${base(instanceId)}/boards/${enc(boardSlug)}/tasks`,
        body,
        signal,
      }),
    executeTaskAction: (instanceId, boardSlug, taskId, body, signal) =>
      transport.request<KanbanTask>({
        method: "POST",
        path: `${base(instanceId)}/boards/${enc(boardSlug)}/tasks/${enc(taskId)}/actions`,
        body,
        signal,
      }),
    addComment: async (instanceId, boardSlug, taskId, text, signal) => {
      await transport.request({
        method: "POST",
        path: `${base(instanceId)}/boards/${enc(boardSlug)}/tasks/${enc(taskId)}/comments`,
        body: { text },
        signal,
        parseJson: false,
      });
    },
    listAssignees: (instanceId, boardSlug, signal) =>
      transport.request<KanbanAssigneeListResponse>({
        path: `${base(instanceId)}/boards/${enc(boardSlug)}/assignees`,
        signal,
      }),
    dispatch: (instanceId, boardSlug, body, signal) =>
      transport.request<KanbanDispatchResult>({
        method: "POST",
        path: `${base(instanceId)}/boards/${enc(boardSlug)}/dispatch`,
        body: body ?? { dryRun: false },
        signal,
      }),
  };
}

File: services/runtime/src/api/v1/kanban.py
"""Kanban API — instance-scoped Hermes Kanban facade (PRD v1.7)."""

from __future__ import annotations

from fastapi import APIRouter, Depends, Query, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from api.deps import get_app_settings, get_db_session
from core.config import Settings
from schemas.kanban import (
    CreateKanbanBoardInput,
    CreateKanbanTaskInput,
    KanbanAssigneeListResponse,
    KanbanBoard,
    KanbanBoardListResponse,
    KanbanCapabilities,
    KanbanCommentCreate,
    KanbanDispatchRequest,
    KanbanDispatchResult,
    KanbanTask,
    KanbanTaskActionInput,
    KanbanTaskDetail,
    KanbanTaskListResponse,
)
from services.kanban_service import KanbanService

router = APIRouter(
    prefix="/instances/{instance_id}/kanban",
    tags=["kanban"],
)


@router.get("/capabilities", response_model=KanbanCapabilities)
async def get_kanban_capabilities(
    instance_id: str,
    settings: Settings = Depends(get_app_settings),
    session: AsyncSession = Depends(get_db_session),
) -> KanbanCapabilities:
    return await KanbanService(settings, session).get_capabilities(instance_id)


@router.get("/boards", response_model=KanbanBoardListResponse)
async def list_kanban_boards(
    instance_id: str,
    include_archived: bool = Query(default=False, alias="includeArchived"),
    settings: Settings = Depends(get_app_settings),
    session: AsyncSession = Depends(get_db_session),
) -> KanbanBoardListResponse:
    return await KanbanService(settings, session).list_boards(
        instance_id,
        include_archived=include_archived,
    )


@router.post("/boards", response_model=KanbanBoard, status_code=status.HTTP_201_CREATED)
async def create_kanban_board(
    instance_id: str,
    body: CreateKanbanBoardInput,
    settings: Settings = Depends(get_app_settings),
    session: AsyncSession = Depends(get_db_session),
) -> KanbanBoard:
    return await KanbanService(settings, session).create_board(instance_id, body)


@router.delete("/boards/{board_slug}", status_code=status.HTTP_204_NO_CONTENT, response_class=Response)
async def delete_kanban_board(
    instance_id: str,
    board_slug: str,
    settings: Settings = Depends(get_app_settings),
    session: AsyncSession = Depends(get_db_session),
) -> Response:
    await KanbanService(settings, session).archive_board(instance_id, board_slug)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/boards/{board_slug}/tasks", response_model=KanbanTaskListResponse)
async def list_kanban_tasks(
    instance_id: str,
    board_slug: str,
    status_filter: str | None = Query(default=None, alias="status"),
    assignee: str | None = None,
    tenant: str | None = None,
    include_archived: bool = Query(default=False, alias="includeArchived"),
    settings: Settings = Depends(get_app_settings),
    session: AsyncSession = Depends(get_db_session),
) -> KanbanTaskListResponse:
    return await KanbanService(settings, session).list_tasks(
        instance_id,
        board_slug,
        status=status_filter,
        assignee=assignee,
        tenant=tenant,
        include_archived=include_archived,
    )


@router.post(
    "/boards/{board_slug}/tasks",
    response_model=KanbanTask,
    status_code=status.HTTP_201_CREATED,
)
async def create_kanban_task(
    instance_id: str,
    board_slug: str,
    body: CreateKanbanTaskInput,
    settings: Settings = Depends(get_app_settings),
    session: AsyncSession = Depends(get_db_session),
) -> KanbanTask:
    return await KanbanService(settings, session).create_task(instance_id, board_slug, body)


@router.get("/boards/{board_slug}/tasks/{task_id}", response_model=KanbanTaskDetail)
async def get_kanban_task(
    instance_id: str,
    board_slug: str,
    task_id: str,
    settings: Settings = Depends(get_app_settings),
    session: AsyncSession = Depends(get_db_session),
) -> KanbanTaskDetail:
    return await KanbanService(settings, session).get_task(instance_id, board_slug, task_id)


@router.post("/boards/{board_slug}/tasks/{task_id}/actions", response_model=KanbanTask)
async def execute_kanban_task_action(
    instance_id: str,
    board_slug: str,
    task_id: str,
    body: KanbanTaskActionInput,
    settings: Settings = Depends(get_app_settings),
    session: AsyncSession = Depends(get_db_session),
) -> KanbanTask:
    return await KanbanService(settings, session).execute_action(
        instance_id,
        board_slug,
        task_id,
        body,
    )


@router.post(
    "/boards/{board_slug}/tasks/{task_id}/comments",
    status_code=status.HTTP_204_NO_CONTENT,
    response_class=Response,
)
async def add_kanban_comment(
    instance_id: str,
    board_slug: str,
    task_id: str,
    body: KanbanCommentCreate,
    settings: Settings = Depends(get_app_settings),
    session: AsyncSession = Depends(get_db_session),
) -> Response:
    await KanbanService(settings, session).add_comment(instance_id, board_slug, task_id, body)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/boards/{board_slug}/assignees", response_model=KanbanAssigneeListResponse)
async def list_kanban_assignees(
    instance_id: str,
    board_slug: str,
    settings: Settings = Depends(get_app_settings),
    session: AsyncSession = Depends(get_db_session),
) -> KanbanAssigneeListResponse:
    return await KanbanService(settings, session).list_assignees(instance_id, board_slug)


@router.post("/boards/{board_slug}/dispatch", response_model=KanbanDispatchResult)
async def dispatch_kanban(
    instance_id: str,
    board_slug: str,
    body: KanbanDispatchRequest | None = None,
    settings: Settings = Depends(get_app_settings),
    session: AsyncSession = Depends(get_db_session),
) -> KanbanDispatchResult:
    return await KanbanService(settings, session).dispatch(
        instance_id,
        board_slug,
        body or KanbanDispatchRequest(),
    )
